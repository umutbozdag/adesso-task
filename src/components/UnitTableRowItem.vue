<template>
  <tr
    class="border-b bg-gray-800 border-gray-700 text-gray-300 cursor-pointer hover:bg-gray-700"
    :class="[rowIndex % 2 === 0 ? 'table__row--even' : '']"
    @click="handleRowClick"
  >
    <!-- TODO: get 'id' from header and render based on that -->
    <!-- OR -->
    <!-- TODO: use 'label' props to pass data and use a slot for more advanced things (rendering costs array) -->
    <td scope="row" class="px-6 py-4 font-medium whitespace-nowrap">
      {{ unitDetail.id }}
    </td>
    <td class="px-6 py-4">{{ unitDetail.name }}</td>
    <td class="px-6 py-4">{{ unitDetail.age }}</td>
    <td class="px-6 py-4">
      <template v-if="unitDetail.costs && unitDetail.costs.length">
        <span v-for="(cost, idx) in unitDetail.costs" :key="cost.name">
          {{ cost.name }}: {{ cost.value
          }}<template v-if="idx !== unitDetail.costs.length - 1">, </template>
        </span>
      </template>
      <template v-else> No data </template>
    </td>
  </tr>
</template>

<script setup lang="ts">
import { IUnitDTO } from '@/types/dto/UnitDTO';

interface UnitTableRowItemProps {
  unitDetail: IUnitDTO;
  rowIndex: number;
}

const emit = defineEmits<{
  (e: 'row-click', unit: IUnitDTO): void;
}>();

const props = defineProps<UnitTableRowItemProps>();

const handleRowClick = () => {
  emit('row-click', props.unitDetail);
};
</script>
