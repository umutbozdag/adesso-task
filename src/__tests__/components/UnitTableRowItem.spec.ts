import UnitTableRowItem from '@/components/UnitTableRowItem.vue';
import { VueWrapper, mount } from '@vue/test-utils';
import { expect, it, describe, beforeAll } from 'vitest';
import { ComponentPublicInstance } from 'vue';
import { generateArrayOfUnits } from '../mocks/unit.mocks';

describe('UnitTableRowItem', () => {
  let wrapper: VueWrapper<ComponentPublicInstance & { [key: string]: any }>;
  beforeAll(() => {
    wrapper = mount(UnitTableRowItem, {
      props: {
        rowIndex: 0,
        unitDetail: generateArrayOfUnits(1)[0]
      }
    });
  });

  it('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it(`emits 'row-clicked' event when handleRowClick`, async () => {
    wrapper.vm.handleRowClick();
    const emittedEvents = wrapper.emitted();
    expect(emittedEvents['row-clicked']).toHaveLength(1);
  });
});
