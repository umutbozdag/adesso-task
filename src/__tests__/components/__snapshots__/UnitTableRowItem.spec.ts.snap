// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UnitTableRowItem > renders correctly 1`] = `
"<tr class=\\"border-b bg-gray-800 border-gray-700 text-gray-300 cursor-pointer hover:bg-gray-700 table__row--even\\">
  <td scope=\\"row\\" class=\\"px-6 py-4 font-medium whitespace-nowrap\\">1</td>
  <td class=\\"px-6 py-4\\">Archer</td>
  <td class=\\"px-6 py-4\\">Dark</td>
  <td class=\\"px-6 py-4\\"><span>Wood: 25, </span><span>Gold: 45<!--v-if--></span></td>
</tr>"
`;
